### PR TITLE
Pod Resource API: graduate getAllocatableResources to Beta 

### PR DESCRIPTION
**Pull request by swatisehgal**
PR: 

- https://github.com/kubernetes/kubernetes/pull/105003

Enhancement issue:

- Extend podresources API to report allocatable resources kubernetes/enhancements#2403

KEP:

- https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2403-pod-resources-allocatable-resources

Also fixes:

- https://github.com/kubernetes/kubernetes/issues/102191

### CODE REVIEW

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 1]
```diff
@@ -239,9 +241,23 @@ message ContainerDevices {
...
+
 GetAllocatableResources provides information on resources initially available on the worker node.
 It provides more information than kubelet exports to APIServer.
 
+{{< note >}}
+`GetAllocatableResources` should only be used to evaluate [allocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable)
+resources on a node. If the goal is to evaluate free/unallocated resources it should be used in
+conjunction with the List() endpoint. The result obtained by `GetAllocatableResources` would remain
+the same unless the underlying resources exposed to kubelet change. This happens rarely but when
+it does (e.g. CPUs onlined/offlined, devices added/removed), client is expected to call
```
Comment (sftim) [inline content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:256]: nit:
```suggestion
it does (for example: CPUs taken online or offline; devices added or removed),
then the client is expected to call the
```

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 2]
```diff
@@ -197,6 +197,8 @@ service PodResourcesLister {
 }
 ```
 
+### `List` gRPC endpoint
```
Comment (sftim) [inline content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:200]: I'd set the fragment identifier explicitly:
```suggestion
### `List` gRPC endpoint {#grpc-endpoint-list}
```

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 3]
```diff
@@ -239,9 +241,23 @@ message ContainerDevices {
 }
 ```
 
+### `GetAllocatableResources` gRPC endpoint
```
Comment (sftim) [inline content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:244]: I'd set the fragment identifier explicitly:
```suggestion
###  ### `GetAllocatableResources` gRPC endpoint {#grpc-endpoint-getallocatableresources}
```

**Review by sftim — COMMENTED** [review]
Setting explicit fragment identifiers leaves room for us to change the heading text in the future whilst preserving existing hyperlinks.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md (hunk 1 above)
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: Ack, Thanks. Updated

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md (hunk 2 above)
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: Ack.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md (hunk 3 above)
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: Ack

**Comment by jlbutler** [comment]
@swatisehgal can you rebase and squash your addition review commits? otherwise looks good to me. 

@sftim any further feedback or reviews needed?

**Comment by swatisehgal** [comment]
> @swatisehgal can you rebase and squash your addition review commits? otherwise looks good to me.
 
Thanks, Rebased and squashed :)

**Comment by swatisehgal** [comment]
@sftim Just want to make sure you are happy with this PR. Let me know if there is anything else you need me to do to unblock this. Thanks!

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 4]
```diff
@@ -259,6 +284,12 @@ message AllocatableResourcesResponse {
 }
 
 ```
+Starting from v1.23, the `GetAllocatableResources` is enabled by default through
```
Comment (sftim) [inline content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:287]: nit:
```suggestion
Starting from Kubernetes v1.23, the `GetAllocatableResources` endpoint is enabled by default.
You can disable it by turning off the
```

**Comment by swatisehgal** [comment]
@sftim A gentle request for another round of review on this PR. Thanks.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 5]
```diff
@@ -246,10 +248,33 @@ message ContainerDevices {
     TopologyInfo topology = 3;
 }
 ```
+``{{< note >}}
```
Comment (kbhawkey) [inline content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:251]: nit: Do you need the two quotes before the note callout?

Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: @kbhawkey Removed, Thanks

**Comment by sftim** [comment]
If this gets LGTM I'm happy for this to merge.

**Comment by swatisehgal** [comment]
/cc @fromanirh @klueska

for node review

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 6]
```diff
@@ -247,9 +249,23 @@ message ContainerDevices {
...
+
 GetAllocatableResources provides information on resources initially available on the worker node.
 It provides more information than kubelet exports to APIServer.
 
+{{< note >}}
+`GetAllocatableResources` should only be used to evaluate [allocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable)
+resources on a node. If the goal is to evaluate free/unallocated resources it should be used in
+conjunction with the List() endpoint. The result obtained by `GetAllocatableResources` would remain
+the same unless the underlying resources exposed to kubelet change. This happens rarely but when
+it does (for example: CPUs taken online or offline; devices added or removed), client is expected to call
```
Comment (fromanirh) [inline content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:273]: this is correct for device changes: changes to device availability (hotplug/hotunplug, device health changes) should be reflected to the allocatability as reported by the endpoint; for cpu and memory calling again the endpoint is not sufficient, the kubelet must be restarted (!!). This is a more general kubelet limitation, but still.

I'd just rephrase the sentence between parens ().

**Review by fromanirh — COMMENTED** [review]
we only need a clarification about reacting to cpu and memory changes, everything else LGTM

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md (hunk 6 above)
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:273]: Good point, will fix this.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 7]
```diff
@@ -247,9 +249,25 @@ message ContainerDevices {
...
 It provides more information than kubelet exports to APIServer.
 
+{{< note >}}
+`GetAllocatableResources` should only be used to evaluate [allocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable)
+resources on a node. If the goal is to evaluate free/unallocated resources it should be used in
+conjunction with the List() endpoint. The result obtained by `GetAllocatableResources` would remain
+the same unless the underlying resources exposed to kubelet change. This happens rarely but when
+it does (for example: hotplug/hotunplug, device health changes), client is expected to call
+`GetAlloctableResources` endpoint.
+However, calling GetAllocatableResources` endpoint is not sufficient in case of cpu and/or memory
```
Comment (fromanirh) [inline content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:275]: it seems a typo sneaked in. Note lack of -> ` <- preceding `GetAllocatableResources`

**Review by fromanirh — COMMENTED** [review]
looks good, but a typo sneaked in, messing up the formatting :\

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md (hunk 7 above)
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:275]: Ah, good catch! Fixing now

Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: Fixed!

**Review by fromanirh — COMMENTED** [review]
/lgtm
from sig-node perspective
thanks for the quick fixes!

**Pull request merged**